monai/data/image_writer.py

```diff
@@ -0,0 +1,395 @@
+
+        - ``set_data_array()`` to set the data array (input must be numpy array or torch tensor)
+            - this method sets the backend object's data part
+        - ``set_metadata()`` to set the metadata and output affine
+            - this method sets the metadata including affine handling and image resampling
+        - backend-specific data object ``create_backend_obj()``
+        - backend-specific writing function ``write()``
+
+    The primary usage of subclasses of ``ImageWriter`` is:
+
+    .. code-block:: python
+
+        writer = MyWriter()  # subclass of ImageWriter
+        writer.set_data_array(data_array)
+        writer.set_metadata(meta_dict)
+        writer.write(filename)
+
+    This creates an image writer object based on ``data_array`` and ``meta_dict`` and write to ``filename``.
+
+    It supports up to three spatial dimensions (with the resampling step supports for both 2D and 3D).
+    When saving multiple time steps or multiple channels `data_array`, time
+    and/or modality axes should be the at the `channel_dim`. For example,
+    the shape of a 2D eight-class and ``channel_dim=0``, the segmentation
+    probabilities to be saved could be `(8, 64, 64)`; in this case
+    ``data_array`` will be converted to `(64, 64, 1, 8)` (the third
+    dimension is reserved as a spatial dimension).
+
+    The ``metadata`` could optionally have the following keys:
+
+        - ``'original_affine'``: for data original affine, it will be the
+            affine of the output object, defaulting to an identity matrix.
+        - ``'affine'``: it should specify the current data affine, defaulting to an identity matrix.
+        - ``'spatial_shape'``: for data output spatial shape.
+
+    When ``metadata`` is specified, the saver will may resample data from the space defined by
+    `"affine"` to the space defined by `"original_affine"`, for more details, please refer to the
+    ``resample_if_needed`` method.
+    """
+
+    def __init__(self, **kwargs):
+        """
+        The constructor supports adding new instance members.
+        The current member in the base class is ``self.data_obj``, the subclasses can add more members,
+        so that necessary meta information can be stored in the object and shared among the class methods.
+        """
+        self.data_obj = None
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+
+    def set_data_array(self, data_array, **kwargs):
+        raise NotImplementedError(f"Subclasses of {self.__class__.__name__} must implement this method.")
+
+    def set_metadata(self, meta_dict: Optional[Mapping], **options):
+        raise NotImplementedError(f"Subclasses of {self.__class__.__name__} must implement this method.")
+
+    def write(self, filename: PathLike, verbose: bool = True, **kwargs):
+        """subclass should implement this method to call the backend-specific writing APIs."""
+        if verbose:
+            logger.info(f"writing: {filename}")
```

Comment:
Seems the `logger` is not defined? Could you please help confirm it?

Thanks.

Reply:
it's a module-level variable, now it's added to `__all__`